server/src/main/java/io/crate/auth/JWTAuthenticationMethod.java

```diff
@@ -0,0 +1,178 @@
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.security.PublicKey;
+import java.security.interfaces.RSAPublicKey;
+import java.time.Clock;
+import java.util.Locale;
+import java.util.concurrent.TimeUnit;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import com.auth0.jwk.Jwk;
+import com.auth0.jwk.JwkException;
+import com.auth0.jwk.JwkProvider;
+import com.auth0.jwk.JwkProviderBuilder;
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.JWTVerifier;
+import com.auth0.jwt.algorithms.Algorithm;
+import com.auth0.jwt.interfaces.DecodedJWT;
+
+import io.crate.protocols.postgres.ConnectionProperties;
+import io.crate.role.Role;
+import io.crate.role.Roles;
+
+public class JWTAuthenticationMethod implements AuthenticationMethod {
+
+    public static final String NAME = "jwt";
+
+    private final Roles roles;
+
+    public JWTAuthenticationMethod(Roles roles) {
+        this.roles = roles;
+    }
+
+    @Override
+    public @Nullable Role authenticate(Credentials credentials, ConnectionProperties connProperties) {
+        var username = credentials.username();
+        assert username != null : "User name must be resolved before authentication attempt";
+        var decodedJWT = credentials.decodedToken();
+        assert decodedJWT != null : "Token must be not null on jwt auth";
+
+        Role user = roles.findUser(username);
+        if (user == null) {
+            throw new RuntimeException("jwt authentication failed for user \"" + username + "\"");
+        }
+
+        try {
+            JwkProvider jwkProvider = jwkProvider(decodedJWT);
+            Algorithm algorithm = resolveAlgorithm(jwkProvider, decodedJWT);
+
+            var jwtProperties = user.jwtProperties();
+            assert jwtProperties != null : "Resolved user must have jwt properties";
+            // Expiration date is checked by default(if provided in token)
+            JWTVerifier.BaseVerification verification = (JWTVerifier.BaseVerification) JWT.require(algorithm)
+                // withers below are not needed for payload signature check.
+                // It's an extra step on top of signature verification to double check that user metadata matches token payload.
+                .withIssuer(jwtProperties.iss())
+                .withClaim("username", jwtProperties.username());
+
+            JWTVerifier verifier = verification.build(clock());
+            verifier.verify(decodedJWT);
+
+        } catch (Exception e) {
+            throw new RuntimeException(
+                String.format(
+                    Locale.ENGLISH,
+                    "jwt authentication failed for user %s. Reason: %s",
+                    username,
+                    e.getMessage()
+                )
+            );
+        }
+
+        return user;
+    }
+
+    public Clock clock() {
+        return Clock.systemUTC();
+    }
+
+    public JwkProvider jwkProvider(@NotNull DecodedJWT decodedJWT) {
```

Comment:
Is there really no higher level API than this?

E.g. the python library provides a `client.get_signing_key_from_jwt(token)` method and then you have a `jwt.decode(token, signing_key, algorithms=["RSA256", ..], ...)`

Reply:
Not that I'm aware of. `JwkProviderBuilder` accepts only domain/URL.

Also, we don't have String token here - it's specifically decoded early only once to avoid double decode (point 2 in PR desc, first decode is unsafe decode for lookup).